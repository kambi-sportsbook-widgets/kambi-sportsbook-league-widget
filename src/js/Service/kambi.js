import { offeringModule, statisticsModule } from 'kambi-widget-core-library';

/**
 * Fetches competition event for given filter and criterion identifier.
 * @param {String} filter
 * @param {number} criterionId Criterion identifier
 * @returns {Promise.<Object>}
 */
const getCompetitionEvent = function(filter, criterionId) {
   // modify filter to match only competitions
   const competitionsFilter = (() => {
      const parts = filter.split('/').filter(termKey => !!termKey);

      for (let i = parts.length; i < 4; i++) {
         parts.push('all');
      }

      parts.push('competitions');

      return parts.join('/');
   })();

   // fetch competitions for previously prepared filter
   return offeringModule.getEventsByFilter(competitionsFilter)
      .then((response) => {
         if (!response || !Array.isArray(response.events)) {
            throw new Error('Invalid response from Kambi API');
         }

         // if criterion identifier is not set just find first event which is a competition
         if (Number.isNaN(criterionId)) {
            return response.events.find(ev => ev.event.type === 'ET_COMPETITION');
         }

         // search for event which is a competition and has a betOffer with given criterion identifier
         return response.events.find((ev) => {
            return ev.event.type === 'ET_COMPETITION' &&
               ev.betOffers.find(bo => bo.criterion.id === criterionId);
         });
      })
      .then((event) => {
         if (!event) {
            throw new Error(`Competition not found for filter=${filter} and criterionId=${criterionId}`);
         }

         // following request will respond with all betOffers
         return offeringModule.getEvent(event.event.id);
      })
      .then((event) => {
         if (event === null) {
            throw new Error('Event not found');
         }

         return event;
      });
};

/**
 * Fetches and prepares data for widget.
 * @param {string} filter Events filter
 * @param {number|null} criterionId Criterion identifier
 * @returns {Promise.<object>}
 */
const getData = function(filter, criterionId) {
   criterionId = parseInt(criterionId, 10);

   return Promise.all([
      getCompetitionEvent(filter, criterionId),
      statisticsModule.getLeagueTableStatistics(filter)
   ])
      .then(([event, statistics]) => {
         // don't look for bet offers if criterion identifier is not set
         const betOffers = Number.isNaN(criterionId) ? []
            : event.betOffers.filter(bo => bo.criterion.id === criterionId);

         return {
            event: event,
            betOffers: betOffers,
            statistics: statistics.leagueTableRows.map((row) => {
               row.goalsDifference = row.goalsFor - row.goalsAgainst;
               row.outcomes = betOffers.map(bo => bo.outcomes.find(oc => oc.participantId === row.participantId));
               return row;
            })
         };
      });
};


export default { getData };
