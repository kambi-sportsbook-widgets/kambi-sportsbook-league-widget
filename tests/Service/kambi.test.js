/* eslint-env jest */

import kambi from '../../src/js/Service/kambi';
import { offeringModule, statisticsModule } from 'kambi-widget-core-library';


jest.mock('kambi-widget-core-library', () => ({
   offeringModule: {
      getEventsByFilter: jest.fn(),
      getEvent: jest.fn()
   },
   statisticsModule: {
      getLeagueTableStatistics: jest.fn()
   }
}));

describe('Kambi service', () => {
   beforeEach(() => {
      offeringModule.getEventsByFilter = jest.fn();

      offeringModule.getEvent = jest.fn();

      statisticsModule.getLeagueTableStatistics = jest.fn();

      console.warn = jest.fn((msg) => {
         return msg;
      });
   });

   it('handles Kambi API returning incorrect response for eventsByFilter request', () => {
      offeringModule.getEventsByFilter = jest.fn(() => new Promise(resolve => resolve(null)));

      return kambi.getData('/test/filter')
         .catch(e => expect(e).toMatchSnapshot());
   });

   it('handles Kambi API returning empty events set', () => {
      offeringModule.getEventsByFilter = jest.fn(() => new Promise(resolve => resolve({ events: [] })));

      return kambi.getData('/test/filter', 123)
         .catch(e => expect(e).toMatchSnapshot());
   });

   it('fails to fetch event proper getting title from event third path', () => {
      offeringModule.getEventsByFilter = jest.fn(() => {
         return new Promise((resolve) => {
            resolve({
               events: [
                  {
                     event: {
                        id: 111,
                        path: [
                           {
                              name: 'path1'
                           },
                           {
                              name: 'path2'
                           },
                           {
                              name: 'path3'
                           },
                        ],
                        type: 'ET_COMPETITION'
                     },
                     betOffers: [
                        {
                           criterion: {
                              id: 123
                           }
                        }
                     ]
                  }
               ]
            })
         });
      });

      offeringModule.getEvent = jest.fn(() => {
         return new Promise((resolve, reject) => {
            reject();
         })
      });

      const leagueTableStatisticsMock = {
         leagueTableRows: [
            {
               goalsFor: 100,
               goalsAgainst: 200
            }
         ]
      };

      statisticsModule.getLeagueTableStatistics = jest.fn((filter) => {
         expect(filter).toEqual('/test/filter');
         return new Promise(resolve => resolve(leagueTableStatisticsMock));
      });

      return kambi.getData('/test/filter', 123)
         .then((data) => {
            expect(data).toMatchSnapshot();
            expect(offeringModule.getEventsByFilter).toHaveBeenCalledTimes(1);
            expect(statisticsModule.getLeagueTableStatistics).toHaveBeenCalledTimes(1);
            expect(offeringModule.getEvent).toHaveBeenCalledTimes(1);
         })
   });


   it('handles Kambi API returning empty response for getEvent request', () => {
      const eventMock = { event: { id: 100, type: 'ET_COMPETITION' } };

      offeringModule.getEventsByFilter = jest.fn(() => new Promise(resolve => resolve({events: [eventMock]})));

      offeringModule.getEvent = jest.fn(eventId => null);

      return kambi.getData('/test/filter')
         .catch(e => expect(e).toMatchSnapshot());
   });

   it('returns correct data when criterionId is given', () => {
      const eventMock = {
         event: { id: 100, type: 'ET_COMPETITION' },
         betOffers: [ {
            criterion: { id: 123 },
            outcomes: [ { participantId: 1000 } ]
         } ]
      };

      const leagueTableStatisticsMock = { leagueTableRows: [ {
         goalsFor: 100,
         goalsAgainst: 200,
         participantId: 1000
      } ] };

      offeringModule.getEventsByFilter = jest.fn((filter) => {
         expect(filter).toEqual('test/filter/all/all/competitions');
         return new Promise(resolve => resolve({events: [eventMock]}));
      });

      offeringModule.getEvent = jest.fn((eventId) => {
         expect(eventId).toEqual(eventMock.event.id);
         return Promise.resolve(eventMock);
      });

      statisticsModule.getLeagueTableStatistics = jest.fn((filter) => {
         expect(filter).toEqual('/test/filter');
         return new Promise(resolve => resolve(leagueTableStatisticsMock));
      });

      return kambi.getData('/test/filter', 123)
         .then(data => {
            expect(data).toMatchSnapshot();
            expect(offeringModule.getEventsByFilter).toHaveBeenCalledTimes(1);
            expect(offeringModule.getEvent).toHaveBeenCalledTimes(1);
            expect(statisticsModule.getLeagueTableStatistics).toHaveBeenCalledTimes(1);
         })
   });

   it('returns correct data when criterionId is not given', () => {
      const eventMock = {
         event: { id: 100, type: 'ET_COMPETITION' }
      };

      const leagueTableStatisticsMock = { leagueTableRows: [ {
         goalsFor: 100,
         goalsAgainst: 200
      } ] };

      offeringModule.getEventsByFilter = jest.fn((filter) => {
         expect(filter).toEqual('test/filter/all/all/competitions');
         return new Promise(resolve => resolve({events: [eventMock]}));
      });

      offeringModule.getEvent = jest.fn((eventId) => {
         expect(eventId).toEqual(eventMock.event.id);
         return eventMock;
      });

      statisticsModule.getLeagueTableStatistics = jest.fn((filter) => {
         expect(filter).toEqual('/test/filter');
         return new Promise(resolve => resolve(leagueTableStatisticsMock));
      });

      return kambi.getData('/test/filter')
         .then(data => {
            expect(data).toMatchSnapshot();
            expect(offeringModule.getEventsByFilter).toHaveBeenCalledTimes(1);
            expect(offeringModule.getEvent).toHaveBeenCalledTimes(0);
            expect(statisticsModule.getLeagueTableStatistics).toHaveBeenCalledTimes(1);
         })
   });

});
