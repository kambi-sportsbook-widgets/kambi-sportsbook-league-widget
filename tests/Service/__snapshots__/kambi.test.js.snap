// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Kambi service fails to fetch event proper getting title from event third path 1`] = `
Object {
  "betOffers": Array [],
  "statistics": Array [
    Object {
      "goalsAgainst": 200,
      "goalsDifference": -100,
      "goalsFor": 100,
      "outcomes": Array [],
    },
  ],
  "title": "path3",
}
`;

exports[`Kambi service handles Kambi API returning empty events set 1`] = `[TypeError: Cannot read property 'leagueTableRows' of undefined]`;

exports[`Kambi service handles Kambi API returning empty response for getEvent request 1`] = `[TypeError: Cannot read property 'leagueTableRows' of undefined]`;

exports[`Kambi service returns correct data when criterionId is given 1`] = `
Object {
  "betOffers": Array [
    Object {
      "criterion": Object {
        "id": 123,
      },
      "outcomes": Array [
        Object {
          "participantId": 1000,
        },
      ],
    },
  ],
  "statistics": Array [
    Object {
      "goalsAgainst": 200,
      "goalsDifference": -100,
      "goalsFor": 100,
      "outcomes": Array [
        Object {
          "participantId": 1000,
        },
      ],
      "participantId": 1000,
    },
  ],
  "title": "",
}
`;

exports[`Kambi service returns correct data when criterionId is not given 1`] = `
Object {
  "betOffers": Array [],
  "statistics": Array [
    Object {
      "goalsAgainst": 200,
      "goalsDifference": -100,
      "goalsFor": 100,
      "outcomes": Array [],
    },
  ],
  "title": "",
}
`;
